mlflow/server/js/src/experiment-tracking/components/ExperimentListView.js

```diff
@@ -1,41 +1,114 @@
 import React, { Component } from 'react';
 import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { debounce } from 'lodash';
+
+import { List } from 'antd';
+import VirtualList from 'rc-virtual-list';
 import {
-  Tree,
   Input,
   Typography,
   CaretDownSquareIcon,
   PlusCircleBorderIcon,
   PencilIcon,
+  Checkbox,
+  SearchIcon,
 } from '@databricks/design-system';
 import { Link, withRouter } from 'react-router-dom';
-import { Experiment } from '../sdk/MlflowMessages';
+import { experimentListSearchInput, searchExperimentsApi } from '../actions';
+import {
+  getExperimentListSearchInput,
+  getExperimentListPreviousSearchInput,
+  getExperiments,
+  getSearchExperimentsNextPageToken,
+  getLoadingMoreExperiments,
+} from '../reducers/Reducers';
 import Routes from '../routes';
 import { CreateExperimentModal } from './modals/CreateExperimentModal';
 import { DeleteExperimentModal } from './modals/DeleteExperimentModal';
 import { RenameExperimentModal } from './modals/RenameExperimentModal';
 import { IconButton } from '../../common/components/IconButton';
+import { Spinner } from '../../common/components/Spinner';
 
 export class ExperimentListView extends Component {
   static propTypes = {
     activeExperimentIds: PropTypes.arrayOf(PropTypes.string).isRequired,
-    experiments: PropTypes.arrayOf(Experiment).isRequired,
     history: PropTypes.object.isRequired,
+    // All below depend on connect
+    dispatchSearchExperimentsApi: PropTypes.func.isRequired,
+    dispatchSearchInput: PropTypes.func.isRequired,
+    experiments: PropTypes.arrayOf(PropTypes.object).isRequired,
+    searchInput: PropTypes.string.isRequired,
+    previousSearchInput: PropTypes.string.isRequired,
+    nextPageToken: PropTypes.string,
+    loadingMore: PropTypes.bool.isRequired,
   };
 
   state = {
     hidden: false,
-    searchInput: '',
     showCreateExperimentModal: false,
     showDeleteExperimentModal: false,
     showRenameExperimentModal: false,
     selectedExperimentId: '0',
     selectedExperimentName: '',
+    checkedKeys: [],
+    innerHeight: 90,
   };
 
+  resizeObserver = null;
+
+  componentDidMount() {
+    // Dynamically set the height of the container list based on root element
+    if (window.ResizeObserver) {
+      this.resizeObserver = new ResizeObserver((entries) => {
+        const height = Math.abs(entries[0].contentRect.height * 0.8);
+        this.setState({ innerHeight: height });
+      });
+
+      this.resizeObserver.observe(document.getElementById('root'));
+    }
```

Comment:
Played with this for a while. It seems like not passing height directly causes [the `useVirtual`](https://github.com/react-component/virtual-list/blob/master/src/List.tsx#L77) check to fail. Which then removes the onscroll listener. So we end up with a nice list of the first 100 experiments, but no event listener to get more.

Tried to hack around it by adding event listeners manually or altering the props in various ways... no success. I can keep cracking at other alternatives. I'd really like to come up with something.

Also tried adding an onscroll to the div/list divs directly, that is horribly non-performant (and bad to use).

Some other options might be:
* Exploring another package: Any opposition? I was trying to not introduce any deps that weren't already transitive.
* Using a manual paginator or load more button: seems like not the best user experience, especially for those with 1000's of experiments. Would really like it to function like any other site someone might visit.

Thanks for the review btw, your feedback has been extremely helpful :).

Reply:
Played around with [react-infinite-scroll-component](https://github.com/ankeetmaini/react-infinite-scroll-component) from [this](https://ant.design/components/list#components-list-demo-infinite-load) antd demo for about a half day yesterday. Doesn't seem to be what we want since adding items to the store causes a re-render of all list items (back to square one). 

Virtual list seems the way to go for this. It seems to work fine (even on resize) with the height set up from the observer. Unless I am missing something.

Reply:
Here's the attempt at infinite scroll. It gets stuck in an infinite loading/scrolling re-rendering loop at about 1500 experiments. Seems like the re-load when props update is causing it to be confused about where it is in the scroll. And it re-renders the entire list on every scroll.

https://github.com/mlflow/mlflow/compare/master...StateFarmIns:mlflow:experiment-pagination-redux-infinite-scroll